R/assess_downloads.r

```diff
@@ -0,0 +1,40 @@
+#' Assess a package for the number of downloads in the past year
+#' 
+#' @details The more times a package has been downloaded the more extensive the user testing and the greater chance there is of someone finding a bug and logging it.
+#'
+#' @eval assess_family_roxygen(
+#'   "downloads_1yr",
+#'   "a numeric value between [0,1] indicating the volume of downloads")
+#'
+#' @export
+assess_downloads_1yr <- function(x, ...){
+  UseMethod("assess_downloads_1yr")
+}
+# assign a friendly name for assess column
+attr(assess_downloads_1yr,"column_name") <- "downloads_1yr"
+attr(assess_downloads_1yr,"label") <- "number of downloads in the past year"
+
+pkg_ref_cache.downloads_1yr <- function(x, name, ...) {
+  cran_downloads(x$name, from=Sys.Date()-365, to=Sys.Date()) %>% group_by(package) %>% summarize(downloads_1yr = sum(count))
+}
+
+#' @import cranlogs
+#' @export
+assess_downloads_1yr.pkg_ref <- function(x, ...) {
+  pkg_metric(x$downloads_1yr, class = "pkg_metric_downloads_1yr")
+}
```

Comment:
If `pkg_ref_cache.downloads_1yr` returns the whole `data.frame`, we could consider the assessment as pulling the sum of the `count` column.

```r
assess_downloads_1yr.pkg_ref <- function(x, ...) {
  downloads <- sum(x$downloads_1yr$count)
  pkg_metric(downloads, class = "pkg_metric_downloads_1yr")
}
```

Reply:
Apologies for not catching this earlier but fields in `x` will need to be named the same as the `pkg_ref_cache` method that is used (e.g. `x$downloads` when a `pkg_ref_cache.downloads` is defined)

```r
pkg_ref_cache.downloads <- function(x, ...) {
  cran_downloads(x$name, from=Sys.Date()-365, to=Sys.Date()) 
}

#' @import cranlogs
#' @export
assess_downloads_1yr.pkg_ref <- function(x, ...) {
  pkg_metric(x$downloads, class = "pkg_metric_downloads_1yr")
}
```

I think this change should resolve the build error

Reply:
I am sorry but I guess I have to ask very stupid questions to get the stuff into the right format. 

If I understand correctly you want the `downloads` and `download_1yr` to be separated, which makes sense so that we could have other assess functions to get 6month, rates etc. 
Does that mean you want the summary of counts (`sum(x$downloads_1yr$count)`) to happen later in the `score` function? I would have expected it in the `assess` function. Does that mean I need another assess function?

Reply:
No worries at all - this is a tricky package, and all work-in-progress code is hard to come back to after time away from the project. 

This is how the code looks now ([permalink](
https://github.com/jmanitz/riskmetric/blob/614d80af132b9bb460ede998e4a426f6d3e2525d/R/assess_downloads.r#L17-L25))
```r
pkg_ref_cache.downloads <- function(x, ...) {
  cran_downloads(x$name, from=Sys.Date()-365, to=Sys.Date()) 
}


#' @import cranlogs
#' @export
assess_downloads_1yr.pkg_ref <- function(x, ...) {
  pkg_metric(x$downloads_1yr, class = "pkg_metric_downloads_1yr")
}
```

When a field in `x` is accessed (like `x$downloads_1yr`), it uses the function `pkg_ref_cache` to try to retrieve it based on the name. The field that you are trying to access has to match the defined `pkg_ref_cache.*` method. 

The problem here is that we're accessing a field called "downloads_1yr", but have defined it as `pkg_ref_cache.downloads`. We just need to change `x$downloads_1yr` to `x$downloads` to match the `pkg_ref_cache` method and we should be good to go.

Reply:
thanks so much for your reply. That part makes sense to me and is an easy fix. However, I think we also need to accumulate the counts in the assess function, right? 

```
pkg_ref_cache.downloads <- function(x, ...) {
  cran_downloads(x$name, from=Sys.Date()-365, to=Sys.Date()) 
}


#' @import cranlogs
#' @export
assess_downloads_1yr.pkg_ref <- function(x, ...) {
  downloads_1yr <- sum(x$downloads$count)
  pkg_metric(downloads_1yr, class = "pkg_metric_downloads_1yr")
}
````

Reply:
Yes, that's right - sorry for the oversight.

Reply:
i committed the changes as discussed, but its still buggy for `list_of_pkg_ref` 
```
> pkg_ref(c("dplyr","ggplot2")) %>% assess_downloads_1yr()
Error in UseMethod("assess_downloads_1yr") : 
  no applicable method for 'assess_downloads_1yr' applied to an object of class "c('list_of_pkg_ref', 'vctrs_list_of', 'vctrs_vctr')"
```